third_party/format_diff/format_diff.py

```diff
@@ -0,0 +1,155 @@
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+#===------------------------------------------------------------------------===#
+"""
+This script reads input from a unified diff and reformats all the changed
+lines. This is useful to reformat all the lines touched by a specific patch.
+Example usage:
+
+  git diff -U0 HEAD^ | python3 format_diff.py yapf -i
+  git diff -U0 HEAD^ | python3 format_diff.py clang-format -i
+
+  svn diff --diff-cmd=diff -x-U0 | python3 format_diff.py clang-format -i -p0
+
+It should be noted that the filename contained in the diff is used unmodified
+to determine the source file to update. Users calling this script directly
+should be careful to ensure that the path in the diff is correct relative to the
+current working directory.
+"""
+
+import argparse
+import difflib
+import io
+import re
+import subprocess
+import sys
+
+BINARY_TO_DEFAULT_REGEX = {
+    "yapf": r".*\.py",
+    "clang-format":
+        r".*\.(cpp|cc|c\+\+|cxx|c|cl|h|hh|hpp|hxx|m|mm|inc|js|ts|proto|"
+        r"protodevel|java|cs)",
+}
+
+
+def parse_arguments():
+  parser = argparse.ArgumentParser(
+      description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
+  parser.add_argument(
+      "binary",
+      help="Location of binary to use for formatting. This controls the "
+      "default values of --regex and --lines-style. If binary isn't 'yapf' "
+      "or 'clang-format' then --regex and --lines-style must be specified.")
+  parser.add_argument(
+      "--regex",
+      metavar="PATTERN",
+      default=None,
+      help="Custom pattern selecting file paths from the diff to reformat. "
+      "Must be specified if --binary is not 'yapf' or 'clang-format', and "
+      "overrides the default --binary sets otherwise. (case sensitive)")
+  parser.add_argument(
+      "--lines-style",
+      default=None,
+      help="How to style the --lines argument for --binary. Must be one of "
+      "Either 'yapf' or 'clang-format'. This must be set manually if --binary "
+      "is not 'yapf' or 'clang-format'.")
+  parser.add_argument(
+      "-p",
+      metavar="NUM",
+      default=1,
+      help="Strip the smallest prefix containing P slashes. Set to 0 if "
+      "passing `--no-prefix` to `git diff` or using `svn`")
+
+  # Parse and error-check arguments
+  args, binary_args = parser.parse_known_args()
+  if args.binary not in BINARY_TO_DEFAULT_REGEX:
+    if not args.regex:
+      raise parser.error("If --binary is not 'yapf' or 'clang-format' then "
+                         "--regex must be set.")
+    if not args.lines_style:
+      raise parser.error("If --binary is not 'yapf' or 'clang-format' then "
+                         "--lines-style must be set.")
+  else:
+    # Set defaults based off of --binary.
+    if not args.regex:
+      args.regex = BINARY_TO_DEFAULT_REGEX[args.binary]
+    if not args.lines_style:
+      args.lines_style = args.binary
```

Comment:
This is likely going to get a bit brittle since the common reason for specifying the binary is a specific version or weird path (like `$HOME/bin/clang-format-9`). That said, we don't need to over generalize this unless it's going to be upstreamed somewhere.

Reply:
We _could_ do something like this:

```python
for key in BINARY_TO_DEFAULT_REGEX:
  if key in args.binary:
    binary = key
    continue
```

This adds the possibility for (unlikely) false positives, and doesn't account for people aliasing `clang-format` to `cf` or something, but it would get rid of most false negatives 🤷‍♀️

Reply:
There's probably some "analytic" way to check if it's a `yapf` or `clang-format` binary too.

Reply:
I'd say punt and follow up with a more robust approach if we can get this into something upstream where it can be shared and community owned.